### PR TITLE
[release/8.0] [Blazor] Close the circuit when all Blazor Server components are disposed

### LINKED ISSUE #48765: Close circuit when there are no remaining interactive Server components
Split out from https://github.com/dotnet/aspnetcore/issues/48396

We don't want to keep the circuit open forever if it's not in use anymore.

Need to decide upon the criteria indicating that a circuit should be closed.

 * Waiting until all server components are gone is an option, but what if a DotNetObjectReference is still in use?
 * Could have a sort of "reference counting" approach and wait for all DotNetObjectReferences to be disposed.
   * But then what do we do about [internal uses](https://github.com/dotnet/aspnetcore/blob/d9b636c35da89dc0bd4fa257754307a882fd1daf/src/Components/Web.JS/src/Rendering/WebRendererInteropMethods.ts#LL17C24-L17C24) of .NET object references? Should we ignore them somehow?

Proposal: just trigger it based on the components only, because this is a UI system, not a general server/client communication channel. Developers who want some kind of headless server/client comms channel can either use SignalR, or they can use `DotNetObjectReference` but have an invisible placeholder component to keep the circuit alive.

### PR DESCRIPTION
**Pull request by MackinnonBuck**
# [Blazor] Close the circuit when all Blazor Server components are disposed

Allows a Blazor Server circuit to close when all root Blazor Server components get dynamically removed from the page.

## Description

The overall approach I've taken is:
1. Define what it means for the circuit to be in use (`WebRootComponentManager.hasAnyExistingOrPendingServerComponents()`):
    * There are interactive Blazor Server components on the page, or...
    * The initialization of an interactive Blazor Server component has started, but hasn't completed yet, or...
    * There are SSR'd components on the page that haven't been initialized for interactivity yet (consider stream rendering, where we don't activate new components until the response completes), but they have either a "Server" or "Auto" render mode
2. Determine the cases where a circuit's "in use" status may have changed:
    * After a render batch is applied (see `attachCircuitAfterRenderCallback` in `WebRootComponentManager.ts`)
      * An applied render batch may result in the creation/disposal of a root component
    * After an SSR update occurs, but before the first render batch is applied
      * Consider the case where an SSR'd component with a Server render mode gets removed from the page, but before the circuit has a chance to initialize
3. Decide what to do if the "in use" status may have changed (`WebRootComponentManager.circuitMayHaveNoRootComponents()`):
    * If the circuit is not in use:
      * Start a timeout with some configurable duration (`SsrStartOptions.circuitInactivityTimeoutMs`), if it hasn't started already
      * When the timeout expires, dispose the circuit
    * If the circuit is not in use:
      * Clear any existing timeout

This PR also:
- [X] Addresses a bug preventing Virtualize from working correctly when a WebAssembly and Server instance is present on the page at the same time
- [X] Adds E2E tests

Fixes #48765

### CODE REVIEW

File: src/Components/Web.JS/src/Services/WebRootComponentManager.ts [hunk 1]
```diff
@@ -135,60 +140,104 @@ export class WebRootComponentManager implements DescriptorHandler, NavigationEnh
...
+  private rootComponentsMayRequireRefresh() {
+    if (this._isComponentRefreshPending) {
+      return;
+    }
+
+    this._isComponentRefreshPending = true;
+    setTimeout(() => {
+      this._isComponentRefreshPending = false;
+      this.refreshRootComponents(this._rootComponents);
+    }, 0);
```
Comment (MackinnonBuck) [inline src/Components/Web.JS/src/Services/WebRootComponentManager.ts:180]: This allows us to liberally call `rootComponentsMayRequireRefresh` without taking the small performance hit from frequent repeated calls to `refreshRootComponents`.

File: src/Components/Web.JS/src/Services/WebRootComponentManager.ts [hunk 2]
```diff
@@ -32,48 +33,52 @@ type RootComponentRemoveOperation = {
...
-  private readonly _rootComponentInfoByDescriptor = new Map<ComponentDescriptor, RootComponentInfo>();
+  private readonly _pendingComponentsToResolve = new Map<number, RootComponentInfo>();
 
-  private _hasPendingRootComponentUpdate = false;
-
-  private _hasStartedCircuit = false;
-
-  private _hasStartedLoadingWebAssembly = false;
-
-  private _hasLoadedWebAssembly = false;
```
Comment (MackinnonBuck) [inline src/Components/Web.JS/src/Services/WebRootComponentManager.ts:53]: These "status" flags were moved to `Boot.Server.Common.ts` and `Boot.WebAssembly.Common.ts` so that there's less of a risk of them getting out of sync with the actual loading/started state of each platform.

File: src/Components/Web.JS/src/Rendering/WebRendererInteropMethods.ts [hunk 3]
```diff
@@ -33,35 +32,26 @@ export function attachWebRendererInterop(
   }
 
   resolveFirstRendererAttached();
-  resolveRendererAttached(rendererId);
+  invokeRendererAttachedListeners(rendererId);
```
Comment (MackinnonBuck) [inline src/Components/Web.JS/src/Rendering/WebRendererInteropMethods.ts:35]: This change is necessary because renders can now attach and detach multiple times.

File: src/Components/Web.JS/src/Platform/SsrStartOptions.ts [hunk 4]
```diff
@@ -7,4 +7,10 @@ export interface SsrStartOptions {
    * (for example, during enhanced navigation or streaming rendering).
    */
   disableDomPreservation?: boolean;
+
+  /**
+   * Configures how long to wait after all Blazor Server components have been removed from the document
+   * before closing the circuit.
+   */
+  circuitInactivityTimeoutMs?: number;
```
Comment (MackinnonBuck) [inline src/Components/Web.JS/src/Platform/SsrStartOptions.ts:15]: We _could_ decide to dispose the circuit as soon as we know the page doesn't contain any server components. However, that can cause async component disposal logic to fail if it involves cleaning up browser-side state. This timeout decides how long components are allowed to take to dispose before the circuit closes. The default is 2 seconds, but that was chosen arbitrarily. We could easily make the default longer or shorter.

File: src/Components/Web.JS/src/Platform/Circuits/CircuitDotNetCallDispatcher.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+
+import { DotNet } from '@microsoft/dotnet-js-interop';
+import { HubConnection } from '@microsoft/signalr';
+
+export class CircuitDotNetCallDispatcher implements DotNet.DotNetCallDispatcher {
+  private _disposed = false;
+
+  constructor(private readonly getCurrentConnection: () => HubConnection) {
+  }
```
Comment (MackinnonBuck) [inline src/Components/Web.JS/src/Platform/Circuits/CircuitDotNetCallDispatcher.ts:11]: This new type allows us to throw an error if the browser initiates a JS interop call to a circuit that's been disposed. This is especially useful for `DotNetObjectReference`s, because you don't want a `DotNetObjectReference` from a disposed circuit to exhibit a "use after free" type of bug where it now points to a random `DotNetObjectReference` from the current circuit.

File: src/Components/Web.JS/src/Boot.Server.Common.ts [hunk 6]
```diff
@@ -119,7 +109,76 @@ export async function startCircuit(components: RootComponentManager<ServerCompon
...
+
+  circuitActive = false;
+
+  await startCircuitPromise;
+
+  if (circuitActive) {
+    // A call to 'startCircuit' was made while we were waiting to dispose the circuit.
+    // Therefore, we should abort the disposal.
+    return;
+  }
```
Comment (MackinnonBuck) [inline src/Components/Web.JS/src/Boot.Server.Common.ts:150]: I appreciate that logic like this can be a bit hard to follow. Unfortunately, we have a recurring problem in Blazor's JS implementation where an assumption was previously made that certain state should be "static" / "singleton" / "module-level", when in fact it's actually better represented as being instance state with a limited lifetime.

In this example, it would be better if there was a more concrete concept of a "circuit" instance on the JavaScript side, where that circuit can be disposed and a new one can be created. This allows a separate copy of state to exist on each circuit instance, so you don't end up with tricky synchronization/ordering challenges during disposal/initialization.

Unfortunately, these kinds of challenges already existed in this file with the disconnection/reconnection logic. It would be possible for us to through and rework this file to use instance-level state (rather than module-level state) at the risk of us potentially changing something subtle in the existing initialization/reconnection logic behavior. Maybe it's worth doing this now, or maybe this is something we should look into more closely for .NET 9.

File: src/Components/Web.JS/src/Platform/Circuits/RenderQueue.ts [hunk 7]
```diff
@@ -20,14 +18,6 @@ export class RenderQueue {
...
   }
 
-  public static getOrCreate(logger: Logger): RenderQueue {
-    if (!RenderQueue.instance) {
-      RenderQueue.instance = new RenderQueue(logger);
-    }
-
-    return this.instance;
-  }
-
```
Comment (MackinnonBuck) [inline src/Components/Web.JS/src/Platform/Circuits/RenderQueue.ts:30]: It doesn't make sense to make `RenderQueue` a singleton anymore because there may be more than once circuit active within the lifetime of the document. It's easier to construct a `RenderQueue` upon circuit initialization rather than having to remember to reset the existing `RenderQueue`'s state when a new circuit gets initialized.

**Comment by SteveSandersonMS** [comment]
> An applied render batch may result in the creation/disposal of a root component

I thought that since render batches (on the client) imply interactive rendering, and interactive components can't contain other interactive root components, this shouldn't be the case. Could you clarify what, if anything, I'm missing? Thanks!

**Comment by MackinnonBuck** [comment]
> I thought that since render batches (on the client) imply interactive rendering, and interactive components can't contain other interactive root components, this shouldn't be the case. Could you clarify what, if anything, I'm missing? Thanks!

A render batch is still sent to the browser for root component disposal, which causes the component to be unregistered as a root component in `BrowserRenderer.ts`. We _could_ close the circuit immediately after the last Server root component gets removed via an SSR update, but this doesn't account for the possibility of dynamically added root components via JavaScript.

File: src/Components/Web.JS/src/Boot.Server.Common.ts [hunk 8]
```diff
@@ -18,42 +18,49 @@ import { sendJSDataStream } from './Platform/Circuits/CircuitStreamingInterop';
...
+let startCircuitPromise: Promise<boolean> | undefined;
 let connection: HubConnection;
 let circuit: CircuitDescriptor;
+let dotNetDispatcher: CircuitDotNetCallDispatcher;
 let dispatcher: DotNet.ICallDispatcher;
-let userOptions: Partial<CircuitStartOptions> | undefined;
+let renderQueue: RenderQueue;
+let options: CircuitStartOptions;
+let logger: ConsoleLogger;
+let afterRenderCallback: (() => void) | undefined;
```
Comment (SteveSandersonMS) [inline src/Components/Web.JS/src/Boot.Server.Common.ts:35]: We're really expanding the amount of module-level state here. Is it at all practical to bundle this together into some object that can be instantiated, as a step towards a future where multiple Blazor instances can coexist?

Comment (SteveSandersonMS) [reply src/Components/Web.JS/src/Boot.Server.Common.ts:35]: Perhaps your other comment at https://github.com/dotnet/aspnetcore/pull/50170/files#r1300507039 already covers this.

File: src/Components/Web.JS/src/Boot.Server.Common.ts (hunk 6 above)
Comment (SteveSandersonMS) [reply src/Components/Web.JS/src/Boot.Server.Common.ts:148]: Yeah, this is tricky. I do think the `Web.JS` code during .NET 8 has gone from "*of mixed quality*" to "*needs a very thorough rewrite if we're going to maintain and expand this long term*". It's always been a lower priority than the C# that runs server-side, and its smallness used to let us get away with idiosyncrasies, but it's no longer small and gets ever more spaghettified.

So I would say:

 * For .NET 8, let's follow your judgement about how much tidying is needed in the short term. We need correctness and to avoid too much weirdness that will make debugging difficult (especially given we may still be patching 8.0 three years from now).
 * For .NET 9, we should be prepared to eat the cost of a bigger renovation of the JS/TS code, even a clean start. I would hope that in something like 3-4 weeks we could go through the process of recreating the functionality with a proper set of Jest tests, copy-pasting *small* chunks from the old code to fit in with a better, instance-oriented architecture.
   * A major asset for us is the E2E tests which are really thorough in most aspects, so we should have good confidence about correctness. That's even more so now a lot of the WebAssembly startup flow is outside this repo, so we no longer have to worry about recreating all the semantics around loading and any errors that can occur.

@MackinnonBuck What do you think? If you think we should be tackling any of this sooner, or if you think a more incremental/evolutionary approach will work better, that's definitely worth discussion.

@mkArtakMSFT What do you think? Would you be prepared to put that much investment in this area in .NET 9? You could think of it as being similar to the big project to align the WebAssembly SDKs - we found time for that in .NET 8 so hopefully we have capacity for a similarly large investment in long term code quality in .NET 9.

**Comment by SteveSandersonMS** [comment]
> A render batch is still sent to the browser for root component disposal, which causes the component to be unregistered as a root component in BrowserRenderer.ts. We could close the circuit immediately after the last Server root component gets removed via an SSR update, but this doesn't account for the possibility of dynamically added root components via JavaScript.

I'm surprised that `WebRootComponentManager` doesn't already know about root components being removed (through all means). Is that not the case?

The only reason I'm asking is that rendering happens thousands of times more often than the set of root components changing, possibly even at 60fps or more, so taking work out of the core render cycle has benefits. Of course, if the amount of work in checking for an empty set of root components is near-zero then it wouldn't matter.

File: src/Components/Web.JS/src/Platform/SsrStartOptions.ts (hunk 4 above)
Comment (SteveSandersonMS) [reply src/Components/Web.JS/src/Platform/SsrStartOptions.ts:15]: This is good. One situation we definitely want to avoid is navigating from one interactive server page to another causing the circuit to be terminated and instantly restarted for the new page. 2 seconds is reasonable.

File: src/Components/Web.JS/src/Services/WebRootComponentManager.ts (hunk 1 above)
Comment (SteveSandersonMS) [reply src/Components/Web.JS/src/Services/WebRootComponentManager.ts:180]: That's good. Would be worth adding a comment for it.

The first time I saw this code I was worried it was some kind of hack (in the bad old days people used to do `setTimeout(..., 0)` as a way to fix order-of-execution issues despite it being impossible reason about) 😄

**Review by SteveSandersonMS — APPROVED** [review]
Looks good!

The sheer amount of changes to how things are wired together is definitely beyond what I can build a mental model for while reading the code, so I definitely can't predict whether there are edge cases where things get stale or similar. Long term I think we need to be prepared for a broader restructure, but I definitely appreciate right now you're making the best choices about working within the existing system.

If there's anything you feel suspicious or uncertain about whether we know it's correct please say so and I'll be happy to help reason about the edge cases.

**Comment by SteveSandersonMS** [comment]
Also this will need to be retargeted to `release/8.0` to go into RC2. It will then be flowed to `main` automatically.

File: src/Components/Web.JS/src/Boot.Server.Common.ts [hunk 9]
```diff
@@ -119,7 +107,89 @@ export async function startCircuit(components: RootComponentManager<ServerCompon
...
+  if (circuitActive) {
+    // A call to 'startCircuit' was made while we were waiting to dispose the circuit.
+    // Therefore, we should abort the disposal.
+    return;
+  }
+
+  // We dispose the .NET dispatcher to prevent it from being used in the future.
+  // This avoids cases where, for example, .NET object references from a
+  // disconnected circuit start pointing to .NET objects for a new circuit.
+  dotNetDispatcher.dispose();
```
[the remark this answers is not in the document]
Comment (javiercn) [reply src/Components/Web.JS/src/Boot.Server.Common.ts:153]: Don't worry folks, the TS39 has you covered https://github.com/tc39/proposal-explicit-resource-management#motivations

File: src/Components/Web.JS/src/Boot.Server.Common.ts (hunk 6 above)
Comment (MackinnonBuck) [reply src/Components/Web.JS/src/Boot.Server.Common.ts]: @SteveSandersonMS I definitely agree that a thorough rework of the JS/TS code is required, and I think this is something we should prioritize early in .NET 9, especially if we intend to do some of the other tracked `blazor.*.js` improvements to support things like microfrontend scenarios, etc.

Regarding circuit management, I've decided to update the implementation now to significantly clean up `Boot.Server.Common.ts`. A lot has been moved to `CircuitManager.ts`, which now logically represents the state and lifetime of a single circuit (whose connection can be closed and reinitialized an arbitrary number of times). While it's a fairly large change, it should be easier to reason about and fix bugs going forward.

File: src/Components/Web.JS/src/Boot.Server.Common.ts (hunk 8 above)
Comment (MackinnonBuck) [reply src/Components/Web.JS/src/Boot.Server.Common.ts]: Yeah, I decided to bite the bullet and give `Boot.Sever.Common.ts` a rework. It should be quite a bit more manageable now.

**Comment by MackinnonBuck** [comment]
Thanks for the review, @SteveSandersonMS!

> I'm surprised that WebRootComponentManager doesn't already know about root components being removed (through all means). Is that not the case?

Unfortunately, it doesn't. It's meant to act as the middleman that keeps in sync SSR'd components and their corresponding interactive components, so it doesn't know anything about root components that get created by other means.

> Of course, if the amount of work in checking for an empty set of root components is near-zero then it wouldn't matter.

In the common case where root components do exist, the impact isn't measurable (at least from my testing with `performance.now()`).
